### ISSUE TITLE
Opensnitch 1.4.0rc1 lets some programs make connections without allowing or denying

### ISSUE THREAD

**Issue by pizzadude** [evilsocket/opensnitch#392]
Hi, I'm using Fedora 33 KDE, and I'm testing Opensnitch 1.4.0rc1. In previous Opensnitch version, if I type "curl http://google.com" in the terminal, opensnitch asks me to allow or deny it. It would deny it if I didn't react to the GUI popup in time. However, in the latest RC of opensnitch, while the GUI is open, the connection still goes through even before I get a chance to allow or deny it. I'm not a programmer, but I think it is caused by this commit:
https://github.com/evilsocket/opensnitch/commit/53bf4f7925885b1d89eec579f760f9d44bb6b689

It defeats the purpose of opensnitch for this to happen.

I also reproduced this with Otter Browser 50% of the time.

**Comment by gustavo-iniguez-goya** [comment]
in this case we're applying the default action. Probably you'll have it configure to Allow everything by default.

Check the preferences or the config:
```
$ cat /etc/opensnitchd/default-config.json
(...) "DefaultAction": "deny", "DefaultDuration": "once"
```

If you configure it to deny, that packet in particular will be denied, and the application will try to send it again. Could you test it?

**Comment by pizzadude** [comment]
For some reason, the default action was "deny" for the UI, but "allow" for the "Nodes". I'm not sure why. After setting action for "Nodes" to "deny", it seems to be working fine.

**Comment by gustavo-iniguez-goya** [comment]
np. Yes, there're 2 different configurations: one for the daemon and one for the GUI.

The idea behind it is that you may have a server with no GUI connected, with a set of rules to allow specific applications/connections, and deny everything by default.

On the other hand, if you run it on a desktop, you can also configure the daemon to block/allow everything by default, but at the same time configure a default action whenever a new process tries to establish a new connection (it'll be applied after the timeout fires). 

I know that it's a bit confusing, but before changing this behaviour, when a pop-up appeared it blocked all the network traffic, causing for example some VPNs to disconnect and some (video) streaming to disconnect and reconnect.

In any case, thank you for reporting it :)

**Comment by pizzadude** [comment]
Yeah, doing this broke my VPN, so I had to switch back. :(

**Comment by gustavo-iniguez-goya** [comment]
actually, if the GUI is connected and a pop-up is asking to allow/deny a new connection, maybe we should apply the GUI's default configuration to the rest of new connections, instead of the daemon's. That would make it behave like pre 1.4.0 for new connections, and at the same time it'd allow already allowed apps/connections.

**Comment by gustavo-iniguez-goya** [comment]
note for myself: the daemon doesn't know the GUI's configuration.

**Comment by raboof** [comment]
> In previous Opensnitch version, if I type "curl http://google.com" in the terminal, opensnitch asks me to allow or deny it. It would deny it if I didn't react to the GUI popup in time. However, in the latest RC of opensnitch, while the GUI is open, the connection still goes through even before I get a chance to allow or deny it

This was surprising to me as well.

> before changing this behaviour, when a pop-up appeared it blocked all the network traffic, causing for example some VPNs to disconnect and some (video) streaming to disconnect and reconnect

I see. I totally agree it is be useful when other traffic can continue while you are still making up your mind about a new one ;)

> if the GUI is connected and a pop-up is asking to allow/deny a new connection, maybe we should apply the GUI's default configuration to the rest of new connections, instead of the daemon's. That would make it behave like pre 1.4.0 for new connections, and at the same time it'd allow already allowed apps/connections. note for myself: the daemon doesn't know the GUI's configuration.

I liked the pre-1.4.0 behavior where (if I understand correctly), the default was 'allow' when the GUI was not yet connected, but effectively switched to 'deny' (drop packets) while the GUI popup was being shown for a new connection.

Could we achieve this again, while also allowing other traffic to continue? Perhaps we could split up the defaults something like:

```
{
    "Server":
    {
        "Address":"unix:///tmp/osui.sock",
        "LogFile":"/var/log/opensnitchd.log"
    },
    "DefaultAction":
    {
        "Headless": "allow",
        "Interactive": "deny"
    }
}
```

... so as soon as the UI connects the default switches from the `Headless` default to the `Interactive` default?

**Comment by pizzadude** [comment]
I downgraded Opensnitch for now because of this issue.

**Comment by BetweenTheKeyboardAndChair** [comment]
As a new user, I was confused by this difference.  I had set the UI to deny, yet items were still going through by default. (The nodes setting was still allow.)

The Windows equivalent that I use is set to deny by default, and it blocks that process (not all processes) until an answer it given.  I find that UI to be much more friendly than the timer.

The timer method means that either I allow by default, or any process that runs for the first time while I am not at the machine gets blocked.  e.g. The system check for updates, and CUPS.

As a new user, the 15s default timer was too short for me to determine the right action before it expired.  I did not see a way to say "give me more time".  I have changed the setting in the UI.  Should the default time be higher to make it more newbie friendly?  (This is a non-issue if it waits for a response if the UI is present.)

**Comment by gustavo-iniguez-goya** [comment]
Yes, you're right. It's confusing and should be fixed. I'll take a look at it for the next release I hope.

> Should the default time be higher to make it more newbie friendly?

Yeah, well, I don't know, that value was the original timeout. Back when the pop-ups blocked all the network traffic, 15s probably was the right value to give you time to act and not to drop network traffic during too much time.

But now that only the connections of one app are stopped, we could increase it, if there's enough demand for it (I don't know what other users think about this).

One trick you can use, is to click on the [+] button of the pop up dialog. That effectively stops the countdown, so you can take your time to analyze the fields of the connection.

**Comment by gustavo-iniguez-goya** [comment]
@pizzadude fixed. Will be included in the next release 1.4.0rc3.

**Comment by pizzadude** [comment]
Thank you :)
